File: 1949.py
import sys


def main():
    sys.stdin = open("inputs/1949.txt", "r")

    T = int(input())

    # 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
    for test_case in range(1, T + 1):
        map_data = []

        N, K = map(int, input().split())

        for x in range(0, N):
            map_data.append(list(map(int, input().split())))

        ret = solution(map_data, N, K)
        print("#%d %d" % (test_case, ret))


def solution(map_data, N, K):
    temp_max = 0

    # find max value
    for y in range(0, N):
        for x in range(0, N):
            if map_data[y][x] > temp_max:
                temp_max = map_data[y][x]

    max_bong_pos = []

    # find max value position
    for y in range(0, N):
        for x in range(0, N):
            if temp_max == map_data[y][x]:
                max_bong_pos.append({'x': x, 'y': y, 'length': -1})

    # find a way from maxbongnam
    for x in range(0, N):
        for y in range(0, N):
            for k in range(0, K + 1):
                map_data[y][x] -= k

                for bong in max_bong_pos:
                    length = move(map_data, bong['x'], bong['y'], N)
                    bong['length'] = length if length > bong['length'] else bong['length']

                map_data[y][x] += k

    max_bong_pos.sort(key=lambda e: e['length'], reverse=True)

    return max_bong_pos[0]['length']


def capable(map_data, x, y, k):
    result = []

    if x > 0 and map_data[y][x - 1] < map_data[y][x]:  # 좌
        result.append(2)
    if y > 0 and map_data[y - 1][x] < map_data[y][x]:  # 상
        result.append(3)
    if x < k - 1 and map_data[y][x + 1] < map_data[y][x]:  # 우
        result.append(0)
    if y < k - 1 and map_data[y + 1][x] < map_data[y][x]:  # 하
        result.append(1)

    return result


def move(map_data, x, y, N):
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]

    way = capable(map_data, x, y, N)
    move_result = 0

    for move_index in way:
        result = move(map_data, x + dx[move_index], y + dy[move_index], N)
        move_result = result if result > move_result else move_result

    return move_result + 1


if __name__ == "__main__":
    main()

File: 1949.cpp
#include <iostream>

#define MAX(a, b) a > b ? a : b
#define MAXN 9

typedef struct bong {
    int x;
    int y;
    int length;
} bong_t;

int map_data[MAXN][MAXN];
bong_t maxbong[100];

using namespace std;

int solution(int N, int K);
void capable(int x, int y, int k);
int move(int x, int y, int N);

int main(int argc, char** argv) {
	freopen("inputs/1949.txt", "r", stdin);

	int T;
	cin >> T;

	for (int test_case = 1; test_case <= T; ++test_case)
	{
        int N, K, ret;
        cin >> N >> K;

        for (int y = 0; y < N; y++) {
            for (int x = 0; x < N; x++) {
                cin >> map_data[y][x];
            }
        }

        ret = solution(N, K);
        cout << "#" << test_case << " " << ret << '\n';
	}

	return 0;//정상종료시 반드시 0을 리턴해야합니다.
}

int solution(int N, int K) {
    int temp_max = 0;
    int maxbong_count = 0;
    int max_length = 0;

    for (int y = 0; y < N; y++)
        for (int x = 0; x < N; x++)
            if (map_data[y][x] > temp_max)
                temp_max = map_data[y][x];

    for (int y = 0; y < N; y++) {
        for (int x = 0; x < N; x++) {
            if (temp_max == map_data[y][x]) {
                maxbong[maxbong_count].x = x;
                maxbong[maxbong_count].y = y;
                maxbong_count++;
            }
        }
    }

    for (int y = 0; y < N; y++) {
        for (int x = 0; x < N; x++) {
            for (int k = 0; k < K + 1; k++) {
                map_data[y][x] -= k;

                for (int b = 0; b < maxbong_count; b++) {
                    maxbong[b].length = 0;

                    int length;
                    length = move(maxbong[b].x, maxbong[b].y, N);
                    max_length = MAX(max_length, length);
                }

                map_data[y][x] += k;
            }
        }
    }

    return max_length;
}

int move(int x, int y, int N) {
    int move_result = 0, wsize = 0;
    int result;

    if (x > 0 && map_data[y][x - 1] < map_data[y][x]) {
        result = move(x - 1, y, N);
        move_result = MAX(move_result, result);
    }
    if (y > 0 && map_data[y - 1][x] < map_data[y][x]) {
        result = move(x, y - 1, N);
        move_result = MAX(move_result, result);
    }
    if (x < N - 1 && map_data[y][x + 1] < map_data[y][x]) {
        result = move(x + 1, y, N);
        move_result = MAX(move_result, result);
    }
    if (y < N - 1 && map_data[y + 1][x] < map_data[y][x]) {
        result = move(x, y + 1, N);
        move_result = MAX(move_result, result);
    }

    return move_result + 1;
}
